File: python/guide/iterator.py
#!/usr/bin/env python3

arr = [1, 2, 3, 4]
it = iter(arr)
print(next(it))
print(next(it))
print(next(it))
print(next(it))

it = iter(arr)
for x in it:
    print(x)


class MyNumbers:
    def __iter__(self):
        self.a = 1
        return self

    def __next__(self):
        if self.a <= 20:
            x = self.a
            self.a += 1
            return x
        else:
            raise StopIteration


myclass = MyNumbers()
myiter = iter(myclass)

for x in myiter:
    print(x)

File: javascript/guide/type/iterator.js
// symbol 数据类型可以作为对象类型的key
console.log(Symbol.iterator);
console.log(Array.prototype[Symbol.iterator]);

const arr = [1, 2, 3];

for (let i of arr) console.log(i);

let iter = arr[Symbol.iterator]();
console.log(typeof iter);
for (let i of iter) console.log(i);

iter = arr[Symbol.iterator]();

console.log(iter[Symbol.iterator]);
console.log(iter.next());
console.log(iter.next());
console.log(iter.next());
console.log(iter.next());
